stablehlo/dialect/VhloOps.cpp

```diff
@@ -176,20 +146,39 @@ VhloDialect::VhloDialect(MLIRContext* context)
 #include "stablehlo/dialect/VhloOps.cpp.inc"
       >();
   addBytecodeInterface(this);
-  addTypes<
-#define GET_TYPEDEF_LIST
-#include "stablehlo/dialect/VhloTypeDefs.cpp.inc"
-      >();
+  addVhloTypes();
   addAttributes<
 #define GET_ATTRDEF_LIST
 #include "stablehlo/dialect/VhloAttrs.cpp.inc"
       >();
 }
 
+void VhloDialect::addVhloTypes() {
+  // Idiomatically, this functionality is expressed as shown below:
+  //
+  //     addTypes<
+  //   #define GET_TYPEDEF_LIST
+  //   #include "stablehlo/dialect/VhloTypeDefs.cpp.inc"
+  //         >();
+  //
+  // However, Dialect::addTypes doesn't work for our situation where we want
+  // to decouple the vhlo_ops and vhlo_types targets. VhloTypeDefs.h.inc only
+  // includes forward declarations of TypeStorage structs, and that's not enough
+  // for Dialect::addTypes to compile.
+  //
+  // Therefore, we work around by introducing this function and then
+  // reimplementing Dialect::addTypes as shown below.
+  addTypesWithoutRegistering<
```

Comment:
Any concerns with this longer term? Is this in-model enough where upstream changes aren't too likely to break it?

Reply:
Time will tell, I guess. If upstream changes break this, we can always revert to the previous state of `vhlo_types` not being self-contained.